cpp/src/generated/parquet_types.cpp

```diff
@@ -8481,11 +8501,27 @@ uint32_t FileMetaData::read(::apache::thrift::protocol::TProtocol* iprot) {
             uint32_t _size326;
             ::apache::thrift::protocol::TType _etype329;
             xfer += iprot->readListBegin(_etype329, _size326);
+            if(read_only_rowgroup_0) {
+              this->row_groups.resize(1);
+            } else {
             this->row_groups.resize(_size326);
+            }
+
             uint32_t _i330;
+            uint32_t rowgroup_size;
             for (_i330 = 0; _i330 < _size326; ++_i330)
             {
-              xfer += this->row_groups[_i330].read(iprot);
+              rowgroup_size = this->row_groups[_i330].read(iprot);
+              xfer += rowgroup_size;
+              if(read_only_rowgroup_0) {
+                break;
+              }
+            }
+            if(read_only_rowgroup_0) {
+              // skip the remaining rowgroups
+              uint32_t skip_len = (_size326 -1) * rowgroup_size;
+              skip_bytes(iprot, skip_len);
+              xfer += skip_len;
```

Comment:
This is probably unsafe since the compact thrift protocol may possibly have a variable number of bytes for each rowgroup. (Though this works in the simple test cases).
I can think of a few other options and would appreciate suggestions:
1. Just exit after reading the first rowgroup and fill out the remaining fields with defaults. (This would be a bummer because then it could not support encryption).
2. Reorder the items in the thrift file so that the (long) rowgroup section comes last.
3. Change the parquet format so that metadata is encoded in a binary rather than a compact format to allow random access / skipping sections.

Reply:
(Background on the compact thrift protocol and why it may give variable size encodings for rowgroups: https://github.com/apache/thrift/blob/master/doc/specs/thrift-compact-protocol.md).